### PR TITLE
<Counter />

### PR DESCRIPTION
**Pull request by fAns1k**

### CODE REVIEW

**Comment by cybergrind** [comment]
Seems like this PR requires rebase

File: README.md [hunk 1]
```diff
@@ -3,6 +3,40 @@ React Native Tipsi custom UI elements
...
+
+handleValueChange = (id) => (console.log(`Current value is ${id}`)
+
+const Example = () => (
+    <Counter
+      step={5}
+      startValue={25}
+      onValueChange={this.handleValueChange}
+    />
+)
```
Comment (itsmepetrov) [inline README.md]: use two spaces instead four

File: README.md [hunk 2]
```diff
@@ -3,6 +3,40 @@ React Native Tipsi custom UI elements
...
+| `startValue` | Initial value of counter. | Number | `1`
+| `onValueChange` | Handle value changes. | Function | `undefined`
+
+#### Example
+
+```js
+import React from 'react'
+import { Counter } from 'tipsi-ui-kit'
+
+handleValueChange = (id) => (console.log(`Current value is ${id}`)
```
Comment (itsmepetrov) [inline README.md]: add `const` before `handleValueChange`

File: README.md [hunk 3]
```diff
@@ -347,13 +381,4 @@ For example let's create `Button` component:
...
-4. Update `uiexplorer/examples` entry file (index.js) to export example for our new component:
-
-    ```js
-    // uiexplorer/examples/index.js
-    import './StarRating'
-    // ...
-    import './Button' // Add this line
-    ```
-
-5. Now you can open `UIExplorer` and click on `<Button />` item to see a result.
```
Comment (itsmepetrov) [inline README.md]: why?

File: __tests__/e2e/11_Counter.test.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    for (let i = 0; i < 3; i += 1) {
+      await driver.click(plus)
+      t.pass(`plus clicked ${i + 1} time`)
+    }
+
+    for (let i = 3; i > 0; i -= 1) {
+      await driver.click(minus)
+      t.pass(`minus clicked ${4 - i} time`)
+    }
+
```
Comment (itsmepetrov) [inline __tests__/e2e/11_Counter.test.js]: We should check if value has changed after the click

File: uiexplorer/examples/Counter.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    description: 'Props: step, startValue',
+    render: () => (
+      <Counter step={5} startValue={25} />
+      ),
+  }, {
+    title: 'Simple example',
+    description: 'Props: onValueChange',
+    render: ({ action }) => ( // eslint-disable-line react/prop-types
+      <Counter onValueChange={action('current value')} />
+      ),
```
Comment (itsmepetrov) [inline uiexplorer/examples/Counter.js]: remove extra spaces

File: uiexplorer/examples/Counter.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import React from 'react'
+import register from '../core/utils/register'
+import Counter from '../../src/Counter'
```
Comment (itsmepetrov) [inline uiexplorer/examples/Counter.js]: import as:

```js
import { Counter } from '../../src'
```

File: src/Counter/Counter.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    startValue: PropTypes.number,
+    step: PropTypes.number,
+    onValueChange: PropTypes.func,
+    styles: PropTypes.object,
+  }
+
+  static defaultProps = {
+    startValue: 1,
+    step: 1,
+    onValueChange: value => value,
```
Comment (itsmepetrov) [inline src/Counter/Counter.js]: change to:

```js
onValueChange: () => {}
```

File: README.md [hunk 8]
```diff
@@ -3,6 +3,40 @@ React Native Tipsi custom UI elements
...
+
+const Example = () => (
+    <Counter
+      step={5}
+      startValue={25}
+      onValueChange={this.handleValueChange}
+    />
+)
+```
+
```
Comment (itsmepetrov) [inline README.md]: Add info about styles structure in documentation

File: README.md (hunk 1 above)
Comment (isnifer) [inline README.md]: hmm, but it is not working example at all, rewrite like this:
```js
import React, { PureComponent } from 'react'
import { Counter } from 'tipsi-ui-kit'

class Example extends PureComponent {
  handleValueChange = id => console.log(`Current value is ${id}`)

  render() {
    <Counter
      step={5}
      startValue={25}
      onValueChange={this.handleValueChange}
    />
  }
}
```

File: README.md [hunk 9]
```diff
@@ -3,6 +3,40 @@ React Native Tipsi custom UI elements
...
+Component to change the number by press "+" or "-".
+
+#### Counter Props
+
+| Name | Desc | Type | Default
+| --- | --- | --- | --- |
+| `step` | Step of counting.  | Number | `1`
+| `startValue` | Initial value of counter. | Number | `1`
+| `onValueChange` | Handle value changes. | Function | `undefined`
```
Comment (isnifer) [inline README.md]: default value for `onValueChange` should be empty function `() => {}`, not *undefined*

File: __tests__/e2e/11_Counter.test.js (hunk 4 above)
Comment (isnifer) [inline __tests__/e2e/11_Counter.test.js]: Expected result of click is changed value, not click itself

File: __tests__/e2e/11_Counter.test.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+      ${counterGroupId}/android.view.View[3]
+    `),
+  })
+
+  try {
+    await helper.openExampleFor('<Counter />')
+
+    await driver.waitForVisible(counter, 20000)
+
+    for (let i = 0; i < 3; i += 1) {
```
Comment (isnifer) [inline __tests__/e2e/11_Counter.test.js]: please define `3` as a constant, like:
```js
const NUMBER_OF_CLICKS = 3

for (let i  = 0; i < NUMBER_OF_CLICKS; i += 1) {
```

File: __tests__/e2e/11_Counter.test.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      await driver.click(plus)
+      t.pass(`plus clicked ${i + 1} time`)
+    }
+
+    for (let i = 3; i > 0; i -= 1) {
```
Comment (isnifer) [inline __tests__/e2e/11_Counter.test.js]: also use it here
```js
// you already define i above 
for (i = 0; i < NUMBER_OF_CLICKS; i += 1) {
```

File: src/Counter/Counter.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    const newCount = this.state.count + this.props.step
+    this.setState({ count: newCount })
+    this.props.onValueChange(newCount)
+  }
+
+  onPressMinus = () => {
+    const newCount = this.state.count - this.props.step
+    this.setState({ count: newCount })
+    this.props.onValueChange(newCount)
+  }
```
Comment (isnifer) [inline src/Counter/Counter.js]: I would like to suggest you rewrite this like:
```js
onPress = (diff) => {
  const count = this.state.count + diff
  this.setState({ count })
  this.props.onValueChange(count)
}

onPressPlus = () => this.props.onPress(this.props.step)

onPressMinus = () => this.props.onPress(-this.props.step)
```

File: __tests__/e2e/11_Counter.test.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    let countText = await driver.waitForVisible(counter, 10000).getText(counter)
+    let count = select({
+      ios: countText[0],
+      android: countText,
+    })
+    t.equal(count, '4', 'count should be 4')
+
+    for (let i = NUMBER_OF_CLICKS; i > 0; i -= 1) {
+      await driver.click(minus)
+      t.pass(`minus clicked ${(NUMBER_OF_CLICKS + 1) - i} time`)
```
Comment (isnifer) [inline __tests__/e2e/11_Counter.test.js]: this construction sooooooo ugly, I suggest you next solution early:
```js
for (i =0; i < NUMBER_OF_CLICKS; i += 1) {
```

File: README.md [hunk 14]
```diff
@@ -3,6 +3,56 @@ React Native Tipsi custom UI elements
...
+### `<Counter />`
+
+Component to change the number by press "+" or "-".
+
+#### Counter Props
+
+| Name | Desc | Type | Default
+| --- | --- | --- | --- |
+| `step` | Step of counting.  | Number | `1`
+| `startValue` | Initial value of counter. | Number | `1`
```
Comment (isnifer) [inline README.md]: Why start value is `1` not `0`?

File: README.md [hunk 15]
```diff
@@ -3,6 +3,56 @@ React Native Tipsi custom UI elements
...
+const handleValueChange = (id) => (console.log(`Current value is ${id}`)
+
+const Example = () => (
+  <Counter
+    step={5}
+    startValue={25}
+    onValueChange={this.handleValueChange}
+  />
+)
+```
```
Comment (isnifer) [inline README.md]: @fAns1k your example has to errors. Did you see my comment https://github.com/tipsi/tipsi-ui-kit/pull/9#discussion_r97569528 ?

File: __tests__/e2e/11_Counter.test.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    countText = await driver
+      .waitForVisible(counter, 10000)
+      .getText(counter)
```
Comment (isnifer) [inline __tests__/e2e/11_Counter.test.js]: 12 lines early you write that code in one line

File: __tests__/e2e/11_Counter.test.js [hunk 17]
```diff
@@ -0,0 +1,84 @@
+import test from 'tape-async'
+import helper from 'tipsi-appium-helper'
+
+const { driver, select, idFromXPath } = helper
+
+test('<Counter />', async (t) => {
+  const counterGroupId = select({
+    ios: idFromXPath(`//
+      XCUIElementTypeScrollView[1]/*/*/
```
Comment (isnifer) [inline __tests__/e2e/11_Counter.test.js]: Please, make selector for an iOS more specific:
```js
ios: idFromXPath(`//
  XCUIElementTypeScrollView[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/
  XCUIElementTypeOther[2]/XCUIElementTypeOther[1]
`),
```

File: __tests__/e2e/11_Counter.test.js [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    const NUMBER_OF_CLICKS = 3
+
+    for (let i = 0; i < NUMBER_OF_CLICKS; i += 1) {
+      await driver.click(plus)
+      t.pass(`plus clicked ${i + 1} time`)
+    }
+
+    let countText = await driver.waitForVisible(counter, 10000).getText(counter)
+    let count = select({
+      ios: countText[0],
```
Comment (isnifer) [inline __tests__/e2e/11_Counter.test.js]: Changes on line 9 will help you remove this bad selector, countText will contain the same result for both platforms

**Comment by isnifer** [comment]
Also test case may be rewritten like this:
```js
const NUMBER_OF_CLICKS = 3
const buttons = [
  { button: plus, expectedValue: '4' },
  { button: minus, expectedValue: '1' },
]

for (const { button, expectedValue } of buttons) {
  for (let i = 0; i < NUMBER_OF_CLICKS; i += 1) {
    await driver.click(button)
  }

  const countText = await driver.waitForVisible(counter, 10000).getText(counter)
  t.equal(countText, expectedValue, `Counter should be changed into ${expectedValue}`)
}
```
It's more clear way - write the same actions for different elements in loops.
Also, this test suite doesn't need info about clicks on button. 
Diff of the counter will indicate that both buttons have clicked.

File: README.md (hunk 14 above)
Comment (fAns1k) [inline README.md]: Why not? If user want another value he can change it.

Comment (isnifer) [inline README.md]: I think, that the most popular case for that is `0`

Comment (fAns1k) [inline README.md]: ok

**Comment by isnifer** [comment]
I think we should provide a property `acceptLessThanZero` that has default value is `false`. // @itsmepetrov

File: src/Counter/Counter.js [hunk 19]
```diff
@@ -0,0 +1,119 @@
+import React, { Component, PropTypes } from 'react'
+import { View, Text, StyleSheet, TouchableOpacity, Platform } from 'react-native'
+import themeable from '../utils/themeable'
+
+class Counter extends Component {
+  static propTypes = {
+    value: PropTypes.number,
```
Comment (isnifer) [inline src/Counter/Counter.js]: You're not define this prop on defaultProps

**Comment by isnifer** [comment]
@itsmepetrov please, review

**Comment by itsmepetrov** [comment]
Looks good to merge

**Pull request merged**